src/java/org/apache/cassandra/db/marshal/VectorType.java

```diff
@@ -0,0 +1,603 @@
+        private final int dimension;
+
+        private Key(AbstractType<?> type, int dimension)
+        {
+            this.type = type;
+            this.dimension = dimension;
+        }
+
+        private VectorType<?> create()
+        {
+            return new VectorType<>(type, dimension);
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Key key = (Key) o;
+            return dimension == key.dimension && Objects.equals(type, key.type);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(type, dimension);
+        }
+    }
+    private static final ConcurrentHashMap<Key, VectorType<?>> instances = new ConcurrentHashMap<>();
+
+    public final AbstractType<T> elementType;
+    public final int dimension;
+    protected final TypeSerializer<T> elementSerializer;
+    private final int valueLengthIfFixed;
+    private final VectorSerializer serializer;
+
+    private VectorType(AbstractType<T> elementType, int dimension)
+    {
+        super(ComparisonType.CUSTOM);
+        if (dimension <= 0)
+            throw new InvalidRequestException(String.format("vectors may only have positive dimentions; given %d", dimension));
+        this.elementType = elementType;
+        this.dimension = dimension;
+        this.elementSerializer = elementType.getSerializer();
+        this.valueLengthIfFixed = elementType.isValueLengthFixed() ?
+                                  elementType.valueLengthIfFixed() * dimension :
+                                  super.valueLengthIfFixed();
+        this.serializer = elementType.isValueLengthFixed() ?
+                          new FixedLengthSerializer() :
+                          new VariableLengthSerializer();
+    }
+
+    public static <T> VectorType<T> getInstance(AbstractType<T> elements, int dimention)
+    {
+        Key key = new Key(elements, dimention);
+        return (VectorType<T>) instances.computeIfAbsent(key, Key::create);
+    }
+
+    @Override
+    public boolean isVector()
+    {
+        return true;
+    }
+
+    @Override
+    public <VL, VR> int compareCustom(VL left, ValueAccessor<VL> accessorL, VR right, ValueAccessor<VR> accessorR)
+    {
+        return getSerializer().compareCustom(left, accessorL, right, accessorR);
+    }
+
+    @Override
+    public int valueLengthIfFixed()
+    {
+        return valueLengthIfFixed;
+    }
+
+    @Override
+    public VectorSerializer getSerializer()
+    {
+        return serializer;
+    }
+
+    public List<ByteBuffer> split(ByteBuffer buffer)
+    {
+        return split(buffer, ByteBufferAccessor.instance);
+    }
+
+    public <V> List<V> split(V buffer, ValueAccessor<V> accessor)
+    {
+        return getSerializer().split(buffer, accessor);
+    }
+
+    public float[] composeAsFloat(ByteBuffer input)
```

Comment:
nit: This method doesn't appear to be used apart from in `VectorTypeTest` where the method is being tested.

Reply:
This was added for @jbellis as he knows his case is always a float, so it avoids the generic overhead